### PR TITLE
Adding Functionality to ES Service Provider

### PR DESCRIPTION
**Pull request by hop-along-polly**

### CODE REVIEW

**Comment by hop-along-polly** [comment]
@blackhatbrigade @josibake @asedarski @krishnaramya please review.

**Comment by asedarski** [comment]
Nothing stands out to me as incorrect syntax but I can't vouch for the functionality :)

**Comment by hop-along-polly** [comment]
@asedarski The heck you can't! That's why the unit tests are there.

File: composer.json [hunk 1]
```diff
@@ -53,7 +53,7 @@
     "require-dev": {
         "phpunit/phpunit": "~4.7",
         "phploc/phploc": "~2.1",
-        "squizlabs/php_codesniffer": "~2.3"
+        "squizlabs/php_codesniffer": "<=2.3"
```
Comment (blackhatbrigade) [inline composer.json:56]: Was this supposed to be less than or equal to?  That doesn't sound good....

File: src/HRQLS/Exceptions/ProtectedIndexException.php [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+/**
+ * Defines a custom exception thrown by the HRQLS framework.
+ *
+ * @package HRQLS\Exceptions
+ */
+ 
+namespace HRQLS\Exceptions;
+ 
+ /**
+  * Defines the custom UsageException class.
```
Comment (blackhatbrigade) [inline src/HRQLS/Exceptions/ProtectedIndexException.php]: Which is used when?  Crap description....

File: src/HRQLS/Exceptions/ProtectedIndexException.php [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  */
+class ProtectedIndexException extends \Exception
```
Comment (blackhatbrigade) [inline src/HRQLS/Exceptions/ProtectedIndexException.php:15]: I love everything about this class though...

File: src/HRQLS/Exceptions/ProtectedIndexException.php (hunk 2 above)
Comment (blackhatbrigade) [inline src/HRQLS/Exceptions/ProtectedIndexException.php]: Wait...  This isn't the usage exceptiong....  YOU LIED TO ME!

File: src/HRQLS/Exceptions/UsageException.php [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+/**
+ * Defines a custom exception thrown by the HRQLS framework.
+ *
+ * @package HRQLS\Exceptions
+ */
+ 
+namespace HRQLS\Exceptions;
+ 
+ /**
+  * Defines the custom UsageException class.
```
Comment (blackhatbrigade) [inline src/HRQLS/Exceptions/UsageException.php:11]: crap crap crap.  I have no idea when this is used based on this description.

File: src/HRQLS/Models/ElasticSearchServiceProvider.php [hunk 5]
```diff
@@ -82,12 +84,188 @@ public function boot(Application $app)
...
+     *      'max_score' => (double), The highest score assigned during execution.
+     *      'hits' => [
+     *      ],
+     *    ],
+     *  ];
      */
-    public function search(array $req)
+    public function search(array $indices, array $types, array $query = [])
     {
+        if (empty($indices)) {
```
Comment (blackhatbrigade) [inline src/HRQLS/Models/ElasticSearchServiceProvider.php:109]: I was going to yell about not setting a default value in the parameter list... but I think this actually looks cleaner.

File: src/HRQLS/Models/ElasticSearchServiceProvider.php [hunk 6]
```diff
@@ -82,12 +84,188 @@ public function boot(Application $app)
     /**
      * Search functionality exposed for use through the service provider.
      *
-     * @param array $req The request to be sent to the ElasticSearch client.
+     * @param array $indices The list of indices to include in the query. Defaults to searching all indices.
+     * @param array $types   The list of types to include in the query. Defaults to searching all types.
+     * @param array $query   The query to run against the specified indexes and types.
```
Comment (blackhatbrigade) [inline src/HRQLS/Models/ElasticSearchServiceProvider.php:89]: I like the new parameter list.  Kinda jealous I didn't think of something like this.

File: src/HRQLS/Models/ElasticSearchServiceProvider.php [hunk 7]
```diff
@@ -82,12 +84,188 @@ public function boot(Application $app)
...
+        
+        if (empty($type)) {
+            $errorMessages[] = '$type cannot be null.';
+        }
+
+        if (empty($doc)) {
+            $errorMessages[] = "An empty document cannot be inserted. That doesn't make any sense.";
+        }
+
+        if (!empty($errorMessages)) {
```
Comment (blackhatbrigade) [inline src/HRQLS/Models/ElasticSearchServiceProvider.php:165]: I like it.  Simple, effective and awesome.

File: src/HRQLS/Models/ElasticSearchServiceProvider.php [hunk 8]
```diff
@@ -82,12 +84,188 @@ public function boot(Application $app)
...
+            throw new UsageException(implode("\n", $errorMessages));
+        }
+
+        $req = [
```
Comment (blackhatbrigade) [inline src/HRQLS/Models/ElasticSearchServiceProvider.php:169]: req?  as in request?  Where'd you learn to name variables?  Visual Studio?

File: src/HRQLS/Models/ElasticSearchServiceProvider.php [hunk 9]
```diff
@@ -82,12 +84,188 @@ public function boot(Application $app)
...
+     * @param string $index   The name of the index the mapping will be added to.
+     * @param string $name    The name of the mapping/type you are creating.
+     * @param array  $mapping An array of mapping properties that define the mapping/type.
+     *
+     * @return array like ['acknowledged' => (Boolean)];
+     *
+     * @throws ProtectedIndexException When $index starts with 'hercules-'.
+     * @throws UsageException When $index, $name, or $mapping are empty/null.
+     */
+    public function addMapping($index, $name, array $mapping)
```
Comment (blackhatbrigade) [inline src/HRQLS/Models/ElasticSearchServiceProvider.php:194]: oh wow... didn't know you were working on this... SUPER awesome...

File: src/HRQLS/Models/ElasticSearchServiceProvider.php [hunk 10]
```diff
@@ -82,12 +84,188 @@ public function boot(Application $app)
...
+       
+        //Attempt to create the index. If the index already exists an indexAlreadyExists Exception is thrown
+        try {
+            $response = $this->client->create($index);
+            if ($response['acknowledged'] != true) {
+                $errorMessages[] = "The request to create {$index} was not acknowledged.";
+            }
+        } catch (\Exception $e) {
+            //Report error message only if it was not an error stating the index already existed.
+            //@TODO It would be better to do this test by type but i'm not sure how to do that in PHP.
```
Comment (blackhatbrigade) [inline src/HRQLS/Models/ElasticSearchServiceProvider.php]: http://bfy.tw/5nMn

**Comment by blackhatbrigade** [comment]
Couple of minor things I'd like to see fixed.  Overall it was great work though!  I'm going to have @asedarski give you a star sticker for doing all of the unit tests.

File: composer.json (hunk 1 above)
Comment (hop-along-polly) [inline composer.json:56]: When I ran composer update there was a PHPCS error about a missing file doc comment in a file that had a file doc comment. I didn't care to figure out why so I did that.

Comment (blackhatbrigade) [inline composer.json:56]: reasonable.  skip it.

File: src/HRQLS/Exceptions/ProtectedIndexException.php (hunk 2 above)
Comment (hop-along-polly) [inline src/HRQLS/Exceptions/ProtectedIndexException.php]: It's called the bait n switch.

File: src/HRQLS/Models/ElasticSearchServiceProvider.php (hunk 8 above)
Comment (hop-along-polly) [inline src/HRQLS/Models/ElasticSearchServiceProvider.php:169]: Do a git blame and you'll see you are the one that started that naming convention in the Service Provider.

File: src/HRQLS/Models/ElasticSearchServiceProvider.php (hunk 10 above)
Comment (hop-along-polly) [inline src/HRQLS/Models/ElasticSearchServiceProvider.php]: I totally deserve that one.

**Comment by asedarski** [comment]
@blackhatbrigade I'll have a Gold star sticker waiting for him ;)

**Comment by hop-along-polly** [comment]
@blackhatbrigade those are some of the best code review comments I've ever seen. I'll get these fixed up tonight ;-)

FYI I love you guys. Thanks for making my night better.

**Comment by hop-along-polly** [comment]
@blackhatbrigade Fixed!

**Pull request merged**